File: intern/1653.minimum-deletions-to-make-string-balanced.js
/*
 * @lc app=leetcode id=1653 lang=javascript
 *
 * [1653] Minimum Deletions to Make String Balanced
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
const minimumDeletions = function (s) {
  const LENGTH = s.length;
  // dpA, dpB 表示在字符串 i 位置以 a， b 结尾的情况下，最少的删除次数
  const dpA = new Array(LENGTH);
  const dpB = new Array(LENGTH);
  if (s[0] == 'a') {
    dpA[0] = 0;
    dpB[0] = 1; //以 b 结尾需要删除1次
  } else {
    dpA[0] = 1;
    dpB[0] = 0;
  }
  for (let i = 1; i < LENGTH; ++i) {
    if (s[i] === 'a') {
      dpA[i] = dpA[i - 1];
      dpB[i] = Math.min(dpA[i - 1], dpB[i - 1]) + 1;
    } else {
      dpA[i] = dpA[i - 1] + 1;
      dpB[i] = Math.min(dpA[i - 1], dpB[i - 1]);
    }
  }
  return Math.min(dpA[LENGTH - 1], dpB[LENGTH - 1]);
};
// @lc code=end

File: src/2023-2024/1653.minimum-deletions-to-make-string-balanced.cpp
/*
 * @lc app=leetcode id=1653 lang=cpp
 *
 * [1653] Minimum Deletions to Make String Balanced
 */

#include <algorithm>
#include <string>
#include <vector>

// @lc code=start
class Solution {
 public:
  int minimumDeletions(const std::string& s) {
    const int N = s.size();
    int previousBCount = 0;
    /*
     * dp[i] - minimum deletions at s[i - 1]
     * */
    std::vector<int> dp(2, 0);

    for (int i = 0; i < N; i++) {
      // At this point, s[0] - s[i] is already a balanced string
      // like aaaaa or aaaabbbb
      if (s[i] == 'b') {
        dp[(i + 1) % 2] = dp[i % 2];
        previousBCount++;
      } else {
        // s[i] == 'a'
        dp[(i + 1) % 2] =
            std::min(dp[i % 2] + 1,  // Delete this `a`
                     previousBCount  // or remove all previous `b`s
            );
      }
    }

    return dp[N % 2];
  }
};
// @lc code=end
